### PR TITLE
No multiple values #49

### PR DESCRIPTION
**Pull request by jgrewe**
Solves issue #49. Removes the "traditional" support for multiple values. The Value entity is gone. Storing multiple values is possible as a list of consistent values, i.e. same type, same unit, same uncertainty.

Values are **always** returned as a list. Even if there is only a single value. This may be discussed and adapted...

Values are serialised as csv using the excel dialect provided by csv, wrapped in brackets.
 
Jason output tests have been disabled for the moment. 

Please check out and try. Before merging we should probably consider the previous pr

### CODE REVIEW

**Comment by lzehl** [comment]
Just for cleaning up the current issue list :)
@jgrewe: with this pr we will also close #38 I guess.
Best Lyuba

**Comment by jgrewe** [comment]
@Luyba, yes I guess that one can be closed then as well :)

@anyone: do not merge yet, it will break a few things and odml version 1 files can't be opened anymore

**Comment by lzehl** [comment]
And, I guess, #12 can be closed as well :)

**Comment by lzehl** [comment]
Hey @jgrewe, I'm going through the changes you made for merging the Value Object into the Property object. This reduction is really nice :) but I have two questions:
1. The previous Value attributes reference can be also just called reference for the Property correct? Currently you define it as value_reference, which I think is unnecessary.
2. The value_str() functions should return the value as string. Why did you implement this? and if the value is a list, this will currently only return the first item of the list as string which is not your plan I guess?

**Comment by jgrewe** [comment]
Hi @lzehl, thanks for the comments. I am happy to rename the function. Regarding the value_str(), good catch! This function is a relict from before. I was not decided whether or not to remove it. In the parser we use str, or unicode to convert individual values. We could either remove the function completely or have it returning the csv representation of the list of values

**Comment by lzehl** [comment]
You're welcome @jgrewe :) in think one could remove the str func completely.

File: odml/dtypes.py [hunk 1]
```diff
@@ -178,43 +180,50 @@ def date_get(string):
...
 
 def tuple_get(string, count=None):
     """
     parse a tuple string like "(1024;768)" and return strings of the elements
     """
-    if not string: return None
+    if not string:
+        return None
     string = string.strip()
     assert string.startswith("(") and string.endswith(")")
```
Comment (achilleas-k) [inline odml/dtypes.py]: Should we be using asserts? (I know this isn't a new line or anything)
Maybe an exception would be more useful.

File: odml/property.py [hunk 2]
```diff
@@ -14,116 +13,178 @@ class Property(base._baseobj):
...
+        >>> p = Property("property1", "a string")
+        >>> p.dtype
+        >>> str
+        >>> p = Property("property1", 2)
+        >>> p.dtype
+        >>> int
+        >>> p = Property("prop", [2, 3, 4])
+        >>> p.dtype
+        >>> int
         :param name: The mane of the property
```
Comment (achilleas-k) [inline odml/property.py]: s/mane/name/

File: odml/property.py [hunk 3]
```diff
@@ -132,46 +193,24 @@ def remove(self, value):
...
-        return self._reorder(self.parent.properties, new_index)
-
-    def __len__(self):
-        return len(self._values)
-
-    def __iter__(self):
-        return self._values.__iter__()
+        if value in self._value:
+            self._value.remove(value)
+        pass
```
Comment (achilleas-k) [inline odml/property.py]: `pass` is redundant

File: odml/tools/xmlparser.py [hunk 4]
```diff
@@ -87,24 +107,26 @@ def save_element(e):
...
                 continue
-
-            if isinstance(val, list):
-                for v in val:
-                    if v is None:
-                        continue
-                    ele = XMLWriter.save_element(v)
-                    cur.append(ele)
+            if isinstance(fmt, format.Property.__class__) and k == "value":
+                # TODO here goes the csv
```
Comment (achilleas-k) [inline odml/tools/xmlparser.py]: Is the TODO done? Should the comment be removed, or is there something missing?

File: odml/tools/xmlparser.py [hunk 5]
```diff
@@ -237,37 +262,40 @@ def parse_tag(self, root, fmt, insert_children=True, create=None):
...
 
         for k, v in arguments.items():
             if hasattr(obj, k):
-                try:
-                    setattr(obj, k, v)
-                except Exception as e:
-                    self.warn("cannot set '%s' property on <%s>: %s" % (k, root.tag, repr(e)), root)
-                    if not self.ignore_errors:
-                        raise
+                if getattr(obj, k) is None:
```
Comment (achilleas-k) [inline odml/tools/xmlparser.py]: Might as well combine the two conditions to
```python
if hasattr(obj, k) and  getattr(obj, k) is None:
```

Or even
```python
if hasattr(obj, k) and obj.k is None
```

File: test/test_property.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+class TestProperty(unittest.TestCase):
+
+    def setUp(self):
+        pass
+
+    def test_value(self):
+        p = Property("property", 100)
+        assert(p.value[0] == 100)
+
+    def test_name(self):
```
Comment (achilleas-k) [inline test/test_property.py]: Should we open an issue to write these tests?

Comment (jgrewe) [inline test/test_property.py]: methinks this is a good plan :)

File: odml/dtypes.py (hunk 1 above)
Comment (jgrewe) [inline odml/dtypes.py]: don't know, in the case string is None, nothing will happen, when we throw an exception, the application will break (if not caught, of course). I would rather not throw an exception.

Comment (achilleas-k) [inline odml/dtypes.py]: I was talking about removing the assert. If string doesn't start and end with `(`, `)`, you'll get an `AssertionError` with no other information.

File: test/test_property.py (hunk 6 above)
Comment (achilleas-k) [inline test/test_property.py]: Added comment to issue #51

**Pull request merged**